optuna/integration/lightgbm_tuner/optimize.py

```diff
@@ -0,0 +1,608 @@
+            elapsed_secs=elapsed_secs,
+            average_iteration_time=average_iteration_time))
+
+        self.trial_count += 1
+
+        return val_score
+
+
+class LightGBMTuner(BaseTuner):
+    """Hyperparameter-tuning with Optuna for LightGBM."""
+
+    def __init__(
+            self,
+            params,  # type: Dict[str, Any]
+            train_set,  # type: lgb.Dataset
+            num_boost_round=1000,  # type: int
+            valid_sets=None,  # type: Optional[VALID_SET_TYPE]
+            valid_names=None,  # type: Optional[Any]
+            fobj=None,  # type: Optional[Callable[Any, Any]]
+            feval=None,  # type: Optional[Callable[Any, Any]]
+            feature_name='auto',  # type: str
+            categorical_feature='auto',  # type: str
+            early_stopping_rounds=None,  # type: Optional[int]
+            evals_result=None,  # type: Optional[Dict[Any, Any]]
+            verbose_eval=True,  # type: Optional[bool]
+            learning_rates=None,  # type: Optional[List[float]]
+            keep_training_booster=False,  # type: Optional[bool]
+            callbacks=None,  # type: Optional[List[Callable[Any, Any]]]
+            time_budget=None,  # type: Optional[int]
+            sample_size=None,  # type: Optional[int]
+            best_params=None,  # type: Optional[Dict[str, Any]]
+            tuning_history=None,  # type: Optional[List[Dict[str, Any]]]
+            enable_adjusting_lr=False,  # type: bool
+            verbosity=1,  # type: Optional[int]  # todo(smly): later.
+    ):
+        params = copy.deepcopy(params)
+        args = [params, train_set]
+        kwargs = dict(num_boost_round=num_boost_round,
+                      valid_sets=valid_sets,
+                      valid_names=valid_names,
+                      fobj=fobj,
+                      feval=feval,
+                      feature_name=feature_name,
+                      categorical_feature=categorical_feature,
+                      early_stopping_rounds=early_stopping_rounds,
+                      evals_result=evals_result,
+                      verbose_eval=verbose_eval,
+                      learning_rates=learning_rates,
+                      keep_training_booster=keep_training_booster,
+                      callbacks=callbacks,
+                      time_budget=time_budget,
+                      verbosity=verbosity,
+                      sample_size=sample_size)  # type: Dict[str, Any]
+        self._parse_args(*args, **kwargs)
+        self.best_booster = None
+
+        self.best_score = -np.inf if self.higher_is_better() else np.inf
+        self.best_params = best_params or {}
+        self.tuning_history = tuning_history or []
+        self.enable_adjusting_lr = enable_adjusting_lr
+
+        if early_stopping_rounds is None:
+            self._suggest_early_stopping_rounds()
+        if valid_sets is None:
+            raise ValueError("`valid_sets` is required.")
+
+    def get_params(self):
```

Comment:
I think `get_params` and some other methods are not supposed to be accessed by users. Please make them private.

Reply:
Fixed: b86b57e